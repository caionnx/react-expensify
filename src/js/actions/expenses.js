import database from '../firebase/firebase'
import modeOperator from './loginModeOperator'

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
})

export const startAddExpense = (expenseData = {}) => (dispatch, getState) => {
  const auth = getState().auth
  const { uid } = auth
  const {
    description = '',
    note = '',
    amount = 0,
    createdAt = 0,
    category = 'none'
  } = expenseData

  const expense = { description, note, amount, createdAt, category }

  return modeOperator(auth, {
    anonymous: () => dispatch(addExpense({
      id: String(Math.floor(Math.random() * 100)),
      ...expense
    })),
    connected: () => database.ref(`users/${uid}/expenses`).push(expense).then(ref => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }))
    })
  })
}

// REMOVE_EXPENSE
export const removeExpense = (id) => ({
  type: 'REMOVE_EXPENSE',
  id
})

export const startRemoveExpense = ({ id } = {}) => (dispatch, getState) => {
  const auth = getState().auth
  const { uid } = auth

  return modeOperator(auth, {
    anonymous: () => dispatch(removeExpense(id)),
    connected: () => database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
      dispatch(removeExpense(id))
    })
  })
}

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})

export const startEditExpense = (id, updates) => (dispatch, getState) => {
  const auth = getState().auth
  const { uid } = auth

  return modeOperator(auth, {
    anonymous: () => dispatch(editExpense(id, updates)),
    connected: () => database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates))
    })
  })
}

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

export const startSetExpenses = () => (dispatch, getState) => {
  const auth = getState().auth
  const { uid } = auth

  return modeOperator(auth, {
    anonymous: () => dispatch(setExpenses([])),
    connected: () => database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
      const expenses = []
      snapshot.forEach(element => {
        expenses.push({
          id: element.key,
          ...element.val()
        })
      })

      dispatch(setExpenses(expenses))
    })
  })
}
