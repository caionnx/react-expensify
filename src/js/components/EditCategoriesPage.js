import React from 'react'
import { connect } from 'react-redux'
import Modal from './Modal'
import slugify from 'slugify'
import PropTypes from 'prop-types'
import {
  startSetCategories,
  startRemoveCategory,
  startAddCategory
} from '../actions/categories'

export class EditCategoriesPage extends React.Component {
  state = {
    removeCategoryModal: false,
    idToRemove: false,
    error: ''
  }
  onOpenRemoveCategoryModal = (id) => {
    this.setState({ removeCategoryModal: true, idToRemove: id })
  }
  onCloseRemoveCategoryModal = () => {
    this.setState({ removeCategoryModal: false })
  }
  onRemove = () => {
    if (this.state.idToRemove) {
      this.props.startRemoveCategory({ id: this.state.idToRemove }).then(() => {
        this.setState({ removeCategoryModal: false })
      })
    }
  }
  onAddCategory = (event) => {
    event.preventDefault()
    const node = event.target.querySelector(`#new-category`)
    const value = node && node.value
    const id = slugify(value, { replacement: '-', remove: null, lower: true })

    if (id === '') {
      this.setState(() => ({ error: 'Please provide category name.' }))
    } else if (this.props.categories.filter((item) => item.id === id).length) {
      this.setState(() => ({ error: 'Category already registered.' }))
    } else {
      this.setState(() => ({ error: '' }))
      this.props.startAddCategory({ id, value })
      node.value = ''
    }
  }
  componentDidMount () {
    const { categories } = this.props

    if (categories && !categories.length) {
      this.props.fillCategories()
    }
  }

  render () {
    const { categories } = this.props

    return (
      <div>
        <div className='page-header'>
          <div className='content-container'>
            <h1 className='page-header__title'>Edit your categories</h1>
          </div>
        </div>
        <div className='content-container'>
          <div className='list-header'>
            <div className='show-for-mobile'>Categories</div>
            <div className='show-for-desktop'>Categories</div>
          </div>
          <div className='list-body'>
            {
              categories && categories.map(
                item =>
                  <div className='list-item' key={item.id}>
                    <h3 className='list-item__title'>{item.value}</h3>
                    <button onClick={() => this.onOpenRemoveCategoryModal(item.id)}>
                      <svg viewBox='0 0 12 12' width='12' height='12'>
                        <path fillRule='evenodd' d='M11.53.47a.75.75 0 0 0-1.061 0l-4.47 4.47L1.529.47A.75.75 0 1 0 .468 1.531l4.47 4.47-4.47 4.47a.75.75 0 1 0 1.061 1.061l4.47-4.47 4.47 4.47a.75.75 0 1 0 1.061-1.061l-4.47-4.47 4.47-4.47a.75.75 0 0 0 0-1.061z' />
                      </svg>
                    </button>
                  </div>
              )
            }
          </div>
          <Modal
            isOpen={this.state.removeCategoryModal}
            label='Remove category'
            className='modal'
            body='Are you sure you want to remove this category?'
            confirmText='Remove'
            cancelText='Cancel'
            thunk
            onConfirm={this.onRemove}
            onClose={this.onCloseRemoveCategoryModal} />

          {this.state.error && <p className='form__error'>{this.state.error}</p>}
          <form onSubmit={this.onAddCategory}>
            <div className='input-group'>
              <div className='input-group__item'>
                <input id='new-category' className='text-input' placeholder='New Category' type='text' />
              </div>
              <div className='input-group__item'>
                <button className='button'>Add Category</button>
              </div>
            </div>
          </form>
        </div>
      </div>
    )
  }
}

EditCategoriesPage.propTypes = {
  categories: PropTypes.array,
  fillCategories: PropTypes.func.isRequired,
  startRemoveCategory: PropTypes.func.isRequired,
  startAddCategory: PropTypes.func.isRequired
}

const mapStateToProps = ({ categories }) => ({
  categories
})

const mapDispatchToProps = (dispatch) => ({
  fillCategories: () => dispatch(startSetCategories()),
  startRemoveCategory: (uid) => dispatch(startRemoveCategory(uid)),
  startAddCategory: (category) => dispatch(startAddCategory(category))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditCategoriesPage)
