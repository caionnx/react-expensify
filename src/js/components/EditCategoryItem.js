import React from 'react'
import PropTypes from 'prop-types'

export const EditCategoryItem = ({ category, buttonAction }) => (
  <div className='list-item list-item--no-link' key={category.id}>
    <h3 className='list-item__title'>{category.value}</h3>
    <button title='Remove category' className='button button--link button--hover-close' onClick={() => buttonAction(category.id)}>
      <svg className='icon' viewBox='0 0 12 12' width='12' height='12'>
        <path fillRule='evenodd' d='M11.53.47a.75.75 0 0 0-1.061 0l-4.47 4.47L1.529.47A.75.75 0 1 0 .468 1.531l4.47 4.47-4.47 4.47a.75.75 0 1 0 1.061 1.061l4.47-4.47 4.47 4.47a.75.75 0 1 0 1.061-1.061l-4.47-4.47 4.47-4.47a.75.75 0 0 0 0-1.061z' />
      </svg>
    </button>
  </div>
)

EditCategoryItem.propTypes = {
  buttonAction: PropTypes.func.isRequired,
  category: PropTypes.object
}

export default EditCategoryItem
