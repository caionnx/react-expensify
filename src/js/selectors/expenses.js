import moment from 'moment'

// Get visible/hidden expenses

export default (expenses, { text, sortBy, startDate, endDate, category }) => {
  const filtered = []
  const unfiltered = []

  expenses.forEach(expense => {
    const createdAtMoment = moment(expense.createdAt)
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())
    const categoryMatch = category && category !== 'none' ? expense.category === category : true

    if (startDateMatch && endDateMatch && textMatch && categoryMatch) {
      filtered.push(expense)
    } else {
      unfiltered.push(expense)
    }
  })

  return {
    filtered: filtered.sort((a, b) => {
      if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1
      } else if (sortBy === 'amount') {
        return a.amount < b.amount ? 1 : -1
      }
    }),
    unfiltered
  }
}
