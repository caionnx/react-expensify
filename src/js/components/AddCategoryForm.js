import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import slugify from 'slugify'
import { startAddCategory } from '../actions/categories'

export class AddCategoryForm extends React.Component {
  state = {
    error: ''
  }

  onAddCategory = (event) => {
    event.preventDefault()
    const node = event.target.querySelector(`#new-category`)
    const value = node && node.value
    const id = slugify(value, { replacement: '-', remove: null, lower: true })

    if (id === '') {
      this.setState(() => ({ error: 'Please provide category name.' }))
      return
    }
    if (this.props.categories.filter((item) => item.id === id).length) {
      this.setState(() => ({ error: 'Category already registered.' }))
      return
    }

    this.setState(() => ({ error: '' }))
    this.props.startAddCategory({ id, value })
    node.value = ''
  }

  render () {
    const { error } = this.state

    return (
      <form onSubmit={this.onAddCategory}>
        {error && <p className='form__error'>{error}</p>}
        <div className='input-group'>
          <div className='input-group__item'>
            <input id='new-category' className='text-input' placeholder='New Category' type='text' />
          </div>
          <div className='input-group__item'>
            <button className='button'>Add Category</button>
          </div>
        </div>
      </form>
    )
  }
}

AddCategoryForm.propTypes = {
  categories: PropTypes.array,
  startAddCategory: PropTypes.func.isRequired
}

const mapStateToProps = ({ categories }) => ({
  categories
})

const mapDispatchToProps = (dispatch) => ({
  startAddCategory: (category) => dispatch(startAddCategory(category))
})

export default connect(mapStateToProps, mapDispatchToProps)(AddCategoryForm)
