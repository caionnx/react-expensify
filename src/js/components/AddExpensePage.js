import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import ExpenseForm from './ExpenseForm'
import { startAddExpense } from '../actions/expenses'

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense).then(() => {
      this.props.history.push('/')
    })
  }
  render () {
    return (
      <div>
        <div className='page-header'>
          <div className='content-container'>
            <h1 className='page-header__title'>Add Expense</h1>
          </div>
        </div>
        <div className='content-container'>
          <ExpenseForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    )
  }
}

AddExpensePage.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }),
  startAddExpense: PropTypes.func.isRequired
}

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage)
