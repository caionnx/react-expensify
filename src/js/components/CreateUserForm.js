import React from 'react'
import { connect } from 'react-redux'
import { startCreateUser } from '../actions/auth'
import PropTypes from 'prop-types'

export class CreateUserForm extends React.Component {
  state = {
    error: false,
    loading: false
  }

  isEmail = (email) =>
    /(.+)@(.+){2,}\.(.+){2,}/.test(email)

  isPassword = (password) =>
    /^(?=.*\d)(?=.*[a-z])(?=.*[A-Z]).{6,8}$/.test(password)

  onFormSubmit = (ev) => {
    ev.preventDefault()
    const formData = new FormData(ev.target) // eslint-disable-line
    const email = formData.get('email')
    const confirmEmail = formData.get('confirm-email')
    const password = formData.get('password')

    if (!this.isEmail(email)) {
      this.setState(() => ({ error: 'Invalid email address.' }))
      return
    }

    if (email !== confirmEmail) {
      this.setState(() => ({ error: 'Please, confirm your email.' }))
      return
    }

    if (!this.isPassword(password)) {
      this.setState(() => ({ error: 'Password must be min 6 characters and max 8 characters. Must include one upper case letter, one lower case letter, and one numeric digit.' }))
      return
    }

    this.setState(() => ({ loading: true }))
    this.props.startCreateUser(email, password).catch((fail) => {
      const error = fail.code === 'auth/email-already-in-use'
        ? fail.message
        : 'Fail to create account.'

      this.setState(() => ({ error, loading: false }))
    })
  }

  render () {
    const { error, loading } = this.state
    const { goBackFunction } = this.props

    return (
      <div>
        <p>Create your account.</p>
        <form className='form' onSubmit={this.onFormSubmit}>
          <input placeholder='Email' aria-label='Email' type='text' name='email' required className='text-input' />
          <input placeholder='Confirm Email' aria-label='Confirm Email' type='text' name='confirm-email' required className='text-input' />
          <input placeholder='Password' aria-label='Password' type='password' name='password' required className='text-input' />
          { error && <p className='form__error'>{error}</p> }
          <button type='submit' className='button' disabled={loading}>
            { loading ? 'Submiting..' : 'Create' }
          </button>
          { goBackFunction &&
            <a href='#' className='button button--secondary' onClick={goBackFunction}>Cancel</a> }
        </form>
      </div>
    )
  }
}

CreateUserForm.propTypes = {
  startCreateUser: PropTypes.func,
  goBackFunction: PropTypes.func
}

const mapDispatchToProps = (dispatch) => ({
  startCreateUser: (emai, password) => dispatch(startCreateUser(emai, password))
})

export default connect(undefined, mapDispatchToProps)(CreateUserForm)
