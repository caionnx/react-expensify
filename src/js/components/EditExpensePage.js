import React from 'react'
import Modal from 'react-modal'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import ExpenseForm from './ExpenseForm'
import { startEditExpense, startRemoveExpense } from '../actions/expenses'

export class EditExpensePage extends React.Component {
  state = {
    removeExpenseModal: false
  }
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense).then(() => {
      this.props.history.push('/dashboard')
    })
  }
  onRemove = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id }).then(() => {
      this.props.history.push('/dashboard')
    })
  }
  onOpenRemoveExpenseModal = () => {
    this.setState({ removeExpenseModal: true })
  }
  onCloseRemoveExpenseModal = () => {
    this.setState({ removeExpenseModal: false })
  }
  componentWillMount () {
    if (!this.props.expense || !this.props.expense.id) {
      this.props.history.push('/dashboard')
    }
  }
  render () {
    return (
      <div>
        <div className='page-header'>
          <div className='content-container'>
            <h1 className='page-header__title'>Edit Expense</h1>
          </div>
        </div>
        <div className='content-container'>
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.onSubmit}
          />
          <button
            id='openRemoveExpenseModal'
            className='button button--secondary'
            onClick={this.onOpenRemoveExpenseModal}>
            Remove Expense
          </button>

          <Modal
            isOpen={this.state.removeExpenseModal}
            onRequestClose={this.onCloseRemoveExpenseModal}
            appElement={document.getElementById('app')}
            contentLabel='Remove Expense Modal'
            closeTimeoutMS={200}
            className='modal'>

            <p className='modal__body'>Are you sure you want to remove this expense?</p>
            <div className='modal__options'>
              <button id='closeRemoveExpenseModal' className='button' onClick={this.onCloseRemoveExpenseModal}>Cancel</button>
              <button id='removeExpense' className='button button--secondary' onClick={this.onRemove}>Remove</button>
            </div>
          </Modal>
        </div>
      </div>
    )
  }
}

EditExpensePage.propTypes = {
  expense: PropTypes.shape({
    id: PropTypes.string
  }),
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }),
  startEditExpense: PropTypes.func.isRequired,
  startRemoveExpense: PropTypes.func.isRequired
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)
})

const mapDispatchToProps = (dispatch, props) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)
