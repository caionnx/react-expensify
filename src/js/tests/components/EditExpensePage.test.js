import React from 'react'
import { shallow } from 'enzyme'
import expenses from '../fixtures/expenses'
import { EditExpensePage, default as ConnectedEditExpensePage } from '../../components/EditExpensePage'
import configureMockStore from 'redux-mock-store'
const mockStore = configureMockStore()

let startEditExpense, startRemoveExpense, history, wrapper

beforeEach(() => {
  startEditExpense = jest.fn(() => Promise.resolve({}))
  startRemoveExpense = jest.fn(() => Promise.resolve({}))
  history = { push: jest.fn() }
  wrapper = shallow(
    <EditExpensePage
      startEditExpense={startEditExpense}
      startRemoveExpense={startRemoveExpense}
      history={history}
      expense={expenses[2]}
    />
  )
})

test('should render EditExpensePage', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should redirect to dashboard if not has expense', () => {
  shallow( // 'mount' from Enzyme causes error
    <EditExpensePage
      startEditExpense={startEditExpense}
      startRemoveExpense={startRemoveExpense}
      history={history}
      expense={{}}
    />
  )

  expect(history.push).toHaveBeenLastCalledWith('/dashboard')
})

test('should handle editExpense', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2])

  expect(startEditExpense).toHaveBeenLastCalledWith(expenses[2].id, expenses[2])

  return startEditExpense().then(() => {
    expect(history.push).toHaveBeenLastCalledWith('/dashboard')
  })
})

test('should open "remove confimation" modal', () => {
  wrapper.find('button#openRemoveExpenseModal').simulate('click')
  const modal = wrapper.find('Modal')
  expect(modal.prop('isOpen')).toBeTruthy()
})

test('should close "remove confimation" modal', () => {
  wrapper.find('button#openRemoveExpenseModal').simulate('click')
  wrapper.find('button#closeRemoveExpenseModal').simulate('click')
  const modal = wrapper.find('Modal')
  expect(modal.prop('isOpen')).toBeFalsy()
})

test('should handle startRemoveExpense', () => {
  wrapper.find('button#removeExpense').simulate('click') // Button inside modal

  expect(startRemoveExpense).toHaveBeenLastCalledWith({
    id: expenses[2].id
  })

  return startRemoveExpense().then(() => {
    expect(history.push).toHaveBeenLastCalledWith('/dashboard')
  })
})

test('should have mapStateToProps and mapDispatchToProps', () => {
  const store = mockStore({ expenses })
  wrapper = shallow(<ConnectedEditExpensePage store={store} match={{params: {id: expenses[0].id}}} />)

  expect(wrapper.prop('expense')).toEqual(expenses[0])
  expect(wrapper.prop('startEditExpense')).toBeInstanceOf(Function)
  expect(wrapper.prop('startRemoveExpense')).toBeInstanceOf(Function)
})
