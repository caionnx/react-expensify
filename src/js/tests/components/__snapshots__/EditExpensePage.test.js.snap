// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 4500,
          "category": "none",
          "createdAt": 48748745645,
          "description": "Credit Card",
          "id": "3",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      id="openRemoveExpenseModal"
      onClick={[Function]}
    >
      Remove Expense
    </button>
    <Modal
      appElement={null}
      ariaHideApp={true}
      bodyOpenClassName="ReactModal__Body--open"
      className="modal"
      closeTimeoutMS={200}
      contentLabel="Remove Expense Modal"
      isOpen={false}
      onRequestClose={[Function]}
      parentSelector={[Function]}
      portalClassName="ReactModalPortal"
      shouldCloseOnEsc={true}
      shouldCloseOnOverlayClick={true}
      shouldFocusAfterRender={true}
      shouldReturnFocusAfterClose={true}
    >
      <p
        className="modal__body"
      >
        Are you sure you want to remove this expense?
      </p>
      <div
        className="modal__options"
      >
        <button
          className="button button--secondary"
          id="removeExpense"
          onClick={[Function]}
        >
          Remove
        </button>
        <button
          className="button"
          id="closeRemoveExpenseModal"
          onClick={[Function]}
        >
          Cancel
        </button>
      </div>
    </Modal>
  </div>
</div>
`;
